src/bidi-engine.js

```diff
@@ -0,0 +1,492 @@
+		var cType = types[idx], wType, nType, len, i;
+		switch ( cType ) {
+			case "L":
+			case "R":
+				_lastArabic = false;
+				break;
+			case "N":
+			case "AN":
+				break;
+
+			case "EN":
+				if ( _lastArabic ) {
+					cType = "AN";
+				}
+				break;
+
+			case "AL":
+				_lastArabic = true;
+				_hasUbatAl = true;
+				cType = "R";
+				break;
+
+			case "WS":
+				cType = "N";
+				break;
+
+			case "CS":
+				if ( idx < 1 || ( idx + 1 ) >= types.length ||
+					( ( wType = resolvedTypes[idx - 1] ) !== "EN" && wType !== "AN" ) ||
+					( ( nType = types[idx + 1] ) !== "EN" && nType !== "AN" ) ) {
+					cType = "N";
+				} else if ( _lastArabic ) {
+					nType = "AN";
+				}
+				cType = ( nType === wType ) ? nType : "N";
+				break;
+
+			case "ES":
+				wType = idx > 0 ? resolvedTypes[idx - 1] : "B";
+				cType = wType === "EN" && ( idx + 1 ) < types.length && types[idx + 1] === "EN" ?
+						"EN" : "N";
+				break;
+
+			case "ET":
+				if ( idx > 0 && resolvedTypes[idx - 1] === "EN" ) {
+					cType = "EN";
+					break;
+				} else if ( _lastArabic ) {
+					cType = "N";
+					break;
+				}
+				i = idx + 1;
+				len = types.length;
+				while ( i < len && types[i] === "ET" ) {
+					i++;
+				}
+				if ( i < len && types[i] === "EN" ) {
+					cType = "EN";
+				} else {
+					cType = "N";
+				}
+				break;
+
+			case "NSM":
+				if ( _isInVisual && !_isInRtl ) { //V->L
+					len = types.length;
+					i = idx + 1;
+					while ( i < len && types[i] === "NSM" ) {
+						i++;
+					}
+					if ( i < len ) {
+						var c = chars[idx];
+						var rtlCandidate = ( c >= 0x0591 && c <= 0x08FF ) || c === 0xFB1E;
+						wType = types[i];
+						if ( rtlCandidate && ( wType === "R" || wType === "AL" ) ) {
+							cType = "R";
+							break;
+						}
+					}
+				}
+				if ( idx < 1 || ( wType = types[idx - 1] ) === "B" ) {
+					cType = "N";
+				} else {
+					cType = resolvedTypes[idx - 1];
+				}
+				break;
+
+			case "B":
+				_lastArabic = false;
+				_hasUbatB = true;
+				cType = _dir;
+				break;
+
+			case "S":
+				_hasUbatS = true;
+				cType = "N";
+				break;
+
+			case "LRE":
+			case "RLE":
+			case "LRO":
+			case "RLO":
+			case "PDF":
+				_lastArabic = false;
+				break;
+			case "BN":
+				cType = "N";
+				break;
+		}
+		return cType;
+	};
+
+	var _handleUbatS = function( types, levels, length ) {
+		for ( var i = 0; i < length; i++ ) {
+			if ( types[i] === "S" ) {
+				levels[i] = _dir;
+				for ( var j = i - 1; j >= 0; j-- ) {
+					if ( types[j] === "WS" ) {
+						levels[j] = _dir;
+					} else {
+						break;
+					}
+				}
+			}
+		}
+	};
+
+	var _invertString = function( text, sourceToTargetMap, levels ) {
+		var charArray = text.split( "" );
+		if ( levels ) {
+			_computeLevels( charArray, levels, { hiLevel: _dir } );
+		}
+		charArray.reverse();
+		sourceToTargetMap && sourceToTargetMap.reverse();
+		return charArray.join( "" );
+	};
+/**
+ * for reference see 3.3 in http://unicode.org/reports/tr9/
+*/
```

Comment:
Some coding styles nitpick... We only use `/**/` style for headers, even long comments use `//` http://contribute.jquery.org/style-guide/js/#comments. Please, could you update these comments using `//` instead. Also indent them according to the below code.